src/client/Microsoft.Identity.Client/ManagedIdentity/ServiceFabricManagedIdentitySource.cs

```diff
@@ -0,0 +1,86 @@
+        {
+        }
+
+        public static ManagedIdentitySource TryCreate(RequestContext requestContext)
+        {
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string identityHeader = EnvironmentVariables.IdentityHeader;
+            string identityServerThumbprint = EnvironmentVariables.IdentityServerThumbprint;
+
+            if (string.IsNullOrEmpty(identityEndpoint) || string.IsNullOrEmpty(identityHeader) || string.IsNullOrEmpty(identityServerThumbprint))
+            {
+                requestContext.Logger.Verbose(() => "[Managed Identity] Service Fabric managed identity unavailable.");
+                return null;
+            }
+
+            if (!Uri.TryCreate(identityEndpoint, UriKind.Absolute, out Uri endpointUri))
+            {
+                throw new MsalClientException(MsalError.InvalidManagedIdentityEndpoint, string.Format(CultureInfo.InvariantCulture, MsalErrorMessage.ManagedIdentityEndpointInvalidUriError, "IDENTITY_ENDPOINT", identityEndpoint, "Service Fabric"));
+            }
+
+            requestContext.Logger.Verbose(() => "[Managed Identity] Creating Service Fabric managed identity. Endpoint URI: " + identityEndpoint);
+            return new ServiceFabricManagedIdentitySource(requestContext, endpointUri, identityHeader);
+        }
+
+        private ServiceFabricManagedIdentitySource(RequestContext requestContext, Uri endpoint, string identityHeaderValue) : base(requestContext)
+        {
+            _endpoint = endpoint;
+            _identityHeaderValue = identityHeaderValue;
+            _userAssignedId = requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedId;
+
+            if (!string.IsNullOrEmpty(_userAssignedId))
+            {
+                requestContext.Logger.Warning(MsalErrorMessage.ManagedIdentityUserAssignedNotConfigurableAtRuntime);
+            }
+        }
+
+        protected override ManagedIdentityRequest CreateRequest(string resource)
+        {
+            ManagedIdentityRequest request = new ManagedIdentityRequest(HttpMethod.Get, _endpoint);
+
+            request.Headers["secret"] = _identityHeaderValue;
+
+            request.QueryParameters["api-version"] = ServiceFabricMsiApiVersion;
+            request.QueryParameters["resource"] = resource;
+
+            if (!string.IsNullOrEmpty(_userAssignedId))
+            {
+                if (Guid.TryParse(_userAssignedId, out _))
+                {
+                    _requestContext.Logger.Info("[Managed Identity] Adding user assigned client id to the request.");
+                    request.QueryParameters[Constants.ManagedIdentityClientId] = _userAssignedId;
+                }
+                else
+                {
+                    _requestContext.Logger.Info("[Managed Identity] Adding user assigned resource id to the request.");
+                    request.QueryParameters[Constants.ManagedIdentityResourceId] = _userAssignedId;
+                }
+            }
```

Comment:
Possible to extract this into a separate method? `ManagedIdentityRequest.AddUserAssignedId(string)` or something.

Reply:
Will be updated with API changes